### ISSUE TITLE
makeinfo errors/warning should be test failures

### ISSUE THREAD

**Issue by cbm755** [catch22/octave-doctest#166]
As @mtmiller mentioned upstream https://savannah.gnu.org/bugs/?49839, `__makeinfo__` make not fail but just errors/warnings to stderr.

How to reproduce:

Suppose I have a typo like `@end exampleXXX` in my docstring, then I get output like:

````
octave:3> doctest vpa
Doctest v0.5.1-dev: this is Free Software without warranty, see source.

/tmp/octave-help-nKwkIO:146: `@end' expected `example', but saw `defun'
vpa .................................................... PASS    3/3   

Summary:

   PASS    3/3   

1/1 targets passed, 0 without tests.
````

That should fail.  I'm not sure we can easily deal with this here: better for upstream needs change to `__makeinfo__`.

**Comment by cbm755** [comment]
See also https://savannah.gnu.org/bugs/?44712 about capturing stderr.

**Comment by cbm755** [comment]
Finally, I glanced at the source of `__makeinfo__`:
````
        cmd = sprintf ("%s --no-headers --no-warn --force --no-validate --output=- %s",
                       makeinfo_program (), name);
````
At least for doctest, we'd probably like to drop that `--no-warn`, although this warning happens even with correctly formatted docstring:
````
/tmp/octave-help-qBtZ2p:114: warning: entry for index `fn' outside of any node
/tmp/octave-help-qBtZ2p:115: warning: entry for index `fn' outside of any node
````

Maybe we should fork a copy of `__makeinfo__`, have it push the stderr to a temp file, load that temp file, filter out this `outside of any node` non-warning.  Something like that could be pushed back upstream too.  @mtmiller, thoughts?

**Comment by mtmiller** [comment]
I'm all for more robust interactions with makeinfo and other tools that Octave relies on. Can we guarantee that we can somehow filter the stderr stream in such a way that if it's non-empty, we can safely treat it as an error that the user / developer can fix? For all supported versions of makeinfo from 4.x through 6.3, in all possible locales? I'm always wary when I see a proposal to parse the output stream of a program into some kind of pass/fail indication.

**Comment by mtmiller** [comment]
@cbm755 if makeinfo sees a real syntax error, it should return a non-zero exit status, which should be propagated and detected by the caller. In your example the error messages comes to the console but it looks like the non-zero exit status is ignored. Is Octave calling `__makeinfo__` in this case or is it doctest?

**Comment by mtmiller** [comment]
If I remove the `--force` option from the makeinfo command line, I get the syntax error, I get a non-zero exit status, and I get no formatted doc string on stdout. With the `--force` option (which Octave uses), it produces the doc string on stdout anyway, still prints the error message, and returns 0.

This option is described as

          --force                 preserve output even if errors.

**Comment by cbm755** [comment]
I see.  So if there was an error, we could call makeinfo again, this time with `--force`, something like:
````
diff -r f04f32f08590 scripts/help/__makeinfo__.m
--- a/scripts/help/__makeinfo__.m       Wed Mar 29 17:02:26 2017 -0700
+++ b/scripts/help/__makeinfo__.m       Wed May 24 07:30:54 2017 -0700
@@ -127,10 +127,10 @@
     ## Take action depending on output type
     switch (lower (output_type))
       case "plain text"
-        cmd = sprintf ("%s --no-headers --no-warn --force --no-validate --output=- %s",
+        cmd = sprintf ("%s --no-headers --no-warn --no-validate --output=- %s",
                        makeinfo_program (), name);
       case "html"
-        cmd = sprintf ("%s --no-headers --html --no-warn --no-validate --force --output=- %s",
+        cmd = sprintf ("%s --no-headers --html --no-warn --no-validate --output=- %s",
                        makeinfo_program (), name);
       otherwise
         error ("__makeinfo__: unsupported output type: '%s'", output_type);
@@ -139,6 +139,15 @@
     ## Call makeinfo
     [status, retval] = system (cmd);
 
+    ## On error, retry with force to ensure user gets something
+    if (status)
+      [status_force, retval] = system ([cmd ' --force']);
+      ## probably returns 0; original return value more useful
+      if (status_force)
+       status = status_force;
+      end
+    end
+
     ## Clean up extra newlines generated by makeinfo
     if (strcmpi (output_type, "plain text"))
       if (numel (retval) > 2 && retval(end-1:end) == "\n\n")
````

**Comment by cbm755** [comment]
posted patch upstream https://savannah.gnu.org/patch/index.php?9464

**Comment by cbm755** [comment]
@mtmiller now that 5.1 is out, could I nudge you to look at this upstream fix?

**Comment by mtmiller** [comment]
This dropped off my radar again, but Rik has now committed the upstream fix for Octave 6.1.

**Comment by cbm755** [comment]
I tried the original example on 6.4.0 and its now:
```
>> doctest vpa
Doctest v0.7.0+: this is Free Software without warranty, see source.

octave-help-Z8njd8:222: `@end' expected `example', but saw `defun'
octave-help-Z8njd8:222: `@end' expected `example', but saw `defun'
vpa .................................................... EXTRACTION ERROR

    __makeinfo__ returned error code 1

```
So closing
